src/Mvc/Mvc.RazorPages/src/ApplicationModels/PageApplicationModel.cs

```diff
@@ -47,6 +48,7 @@ public class PageApplicationModel
             HandlerMethods = new List<PageHandlerModel>();
             HandlerProperties = new List<PagePropertyModel>();
             HandlerTypeAttributes = handlerAttributes;
+            EndpointMetadata = new CopyOnWriteList<object>(ActionDescriptor.EndpointMetadata ?? Array.Empty<object>());
```

Comment:
I don't think this is a good enough reason to change the contract of `CopyOnWriteList`. We use that in a few perf-sensitive places that run every request. This happens once per page.